### PR TITLE
Provide mechanic to check equality of non-value types

### PR DESCRIPTION
**Pull request by blitz-1306**
## Status
This is PR a PoC implementation for #122.

## Changes
- [x] Introduce `eq_encoded()` Scribble builtin to simplify comparing of complex types.
  - Function implementaion is added to Scribble utils contract on-demand (when function is used by any annotation).
  - When any passed argument is not castable to `bytes memory`, it is wrapped with `abi.encode()` (to reduce amount of manual efforts).
- [x] Optimize code of generating utils library (introduce more granular functions).
- [x] Add test sample.

## Notes
Feature discussion goes in #122. See some questions in https://github.com/ConsenSys/scribble/issues/122#issuecomment-1541514034.

Regards.

### CODE REVIEW

File: src/spec-lang/tc/typecheck.ts [hunk 1]
```diff
@@ -1962,97 +1962,124 @@ export function tcForAll(expr: SForAll, ctx: STypingCtx, typeEnv: TypeEnv): Type
...
+                    expr
+                );
+            }
+
+            /**
+             * @todo Consider checking that type is encodable.
+             * Here is how compiler does it:
+             * https://github.com/ethereum/solidity/blob/0a0c389541c0f247691951edd64451be7145436c/libsolidity/ast/Types.cpp#L314-L339
+             */
+            tc(expr.args[0], ctx, typeEnv);
```
Comment (cd1m0) [inline src/spec-lang/tc/typecheck.ts:2020]: Here we need to check that both argument types are 'abi-encodable'. Maybe we can add a new function `isTypeABIEncodable(t: TypeNode): bool` that follows the logic of `toABIEncdodedType`

File: test/samples/eq_encoded.sol [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.19;
```
Comment (cd1m0) [inline test/samples/eq_encoded.sol:2]: Could we add an executable regression test that calls eq_encoded (i.e. a vm test?)?

We should have a negative case (checks that eq_encoded fails when we want it to), and a positive test, for each of the following cases:

1. Comparing two things of different types (e.g. number and array) - negative test, always fails
2. Comparing two primitive things of the same type - numbers, addresses, booleans, fixed bytes, enums (should succeed when equal, fail when not)
3. Comparing two arrays (again check that eq_encoded succeeds when equal, fails when not)
4. Comparing two structs (same 2 tests)

**Review by cd1m0 — CHANGES_REQUESTED** [review]
Good job! Some small tweaks and tests requested

File: test/samples/eq_encoded.sol (hunk 2 above)
Comment (blitz-1306) [reply test/samples/eq_encoded.sol:2]: Will do.

File: src/spec-lang/tc/semcheck.ts [hunk 3]
```diff
@@ -481,6 +481,16 @@ export function scFunctionCall(
             return { isOld: isOld, isConst: allArgsConst, canFail: argsInfo[0].canFail };
         }
 
+        if (callee.name === ScribbleBuiltinFunctions.eq_encoded) {
+            const isOld = ctx.isOld || argsInfo.some((argInfo) => argInfo.isOld);
```
[the remark this answers is not in the document]
Comment (blitz-1306) [reply src/spec-lang/tc/semcheck.ts:485]: Will address.

File: src/instrumenter/utils_library.ts [hunk 4]
```diff
@@ -320,15 +305,214 @@ export function generateUtilsLibrary(
...
+        false,
+        DataLocation.Default,
+        StateVariableVisibility.Default,
+        Mutability.Mutable,
+        "<missing>",
+        undefined,
+        factory.makeElementaryTypeName("uint256", "uint256")
+    );
+
+    const byteCheck = factory.makeForStatement(
```
[the remark this answers is not in the document]
Comment (blitz-1306) [reply src/instrumenter/utils_library.ts:412]: Will do.

File: test/samples/eq_encoded.sol (hunk 2 above)
Comment (blitz-1306) [reply test/samples/eq_encoded.sol:2]: Addressed in e060422

File: src/spec-lang/tc/semcheck.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (blitz-1306) [reply src/spec-lang/tc/semcheck.ts]: Addressed in e060422

File: src/instrumenter/utils_library.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (blitz-1306) [reply src/instrumenter/utils_library.ts]: Addressed in e060422

File: src/spec-lang/tc/typecheck.ts (hunk 1 above)
Comment (blitz-1306) [reply src/spec-lang/tc/typecheck.ts:2020]: See https://github.com/ConsenSys/solc-typed-ast/issues/206

**Review by cd1m0 — APPROVED** [review]
lgtm

**Pull request merged**